### PR TITLE
Feature/tda 83

### PR DESCRIPTION
**Pull request by duprijil**

### CODE REVIEW

File: ted_data_eu/adapters/storage.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        """
+           Implements interaction with ElasticSearch storage by using its API.
+
+            :param index: ElasticSearch index (schema) where the documents will be stored
+            :param host: Elastic API host and port (if different from the defaults for http and https)
+            :param user: Elastic API user
+            :param password: Elastic API password
+            :return:
+        """
+        self.host = host if host else config.ELASTIC_HOST
```
Comment (kaleanych) [inline ted_data_eu/adapters/storage.py:33]: self.host = host or config.ELASTIC_HOST

Also for similar inline ifs

it is more compact

**Review by kaleanych — APPROVED** [review]
Looks ok. Check the comments

File: ted_data_eu/__init__.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 
 PROJECT_RESOURCES_PATH = pathlib.Path(__file__).parent.resolve() / "resources"
 PROJECT_RESOURCES_BQ_FOLDER_PATH = PROJECT_RESOURCES_PATH / "sparql_queries"
+ELASTIC_SUBDOMAIN = 'elastic'
```
Comment (CaptainOfHacks) [inline ted_data_eu/__init__.py:11]: remove this from constants

File: ted_data_eu/__init__.py [hunk 3]
```diff
@@ -37,7 +38,30 @@ def GRAPHDB_HOST(self, config_value: str) -> str:
...
+    def ELASTIC_PASSWORD(self, config_value: str) -> str:
+        return config_value
+
+    @env_property()
+    def ELASTIC_VERSION(self, config_value: str) -> str:
+        return config_value
+
+    @env_property()
+    def ELASTIC_HOST(self, config_value: str) -> str:
+        return f"https://{ELASTIC_SUBDOMAIN}.{self.DOMAIN}"
```
Comment (CaptainOfHacks) [inline ted_data_eu/__init__.py:61]: ELASTIC_HOST need to be a full URL in .env and register in Vault,

File: ted_data_eu/__init__.py [hunk 4]
```diff
@@ -37,7 +38,30 @@ def GRAPHDB_HOST(self, config_value: str) -> str:
         return config_value
 
 
-class TedDataConfigResolver(TedConfigResolver, BQResourcesConfig, GraphDBConfig):
+class ElasticConfig:
+
+    @env_property()
+    def DOMAIN(self, config_value: str) -> str:
```
Comment (CaptainOfHacks) [inline ted_data_eu/__init__.py:44]: remove this secret, is used somewhere?

File: ted_data_eu/adapters/storage.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            :param user: Elastic API user
+            :param password: Elastic API password
+            :return:
+        """
+        self.host = host or config.ELASTIC_HOST
+        self.auth = HTTPBasicAuth(user or config.ELASTIC_USER,
+                                  password or config.ELASTIC_PASSWORD)
+        self.index = index
+
+    def add_document(self, document: str) -> Response:
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:38]: document will be string? or JSON string, if JSON string I recommand to use dict object and convert to JSON string inside adapter.

Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:38]: Result, as Response is very dependent by Elastic behaviour, I recommend checking response and raise exception if something went wrong.

File: ted_data_eu/adapters/storage.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+           Add document to storage.
+
+            :param document: Document to be stored
+            :return:
+        """
+        response = requests.post(url=f"{self.host}/{self.index}/_doc/",
+                                 headers=BASIC_HEADERS,
+                                 auth=self.auth,
+                                 data=document)
+        return response
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:49]: Result, as Response is very dependent by Elastic behaviour, I recommend checking response and raise exception if something went wrong.

File: ted_data_eu/adapters/storage.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+BASIC_HEADERS = {'Content-type': 'application/x-ndjson'}
+
+
+class ElasticStorage(DocumentStorageABC):
+    """
+       Implements interaction with ElasticSearch storage by using its API.
+    """
+
+    def __init__(self,
+                 index: str,
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:20]: I recommend renaming index, in elastic_index

File: ted_data_eu/adapters/storage.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        for document in documents:
+            data_to_send.append(json.dumps({"create": {"_index": self.index}}))
+            data_to_send.append(document)
+        response = requests.post(url=f"{self.host}/{self.index}/_bulk",
+                                 headers=BASIC_HEADERS,
+                                 auth=self.auth,
+                                 data='\n'.join(line for line in data_to_send) + '\n')
+        return response
+
+    def query(self, query: str) -> Response:
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:68]: Result, as Response is very dependent by Elastic behaviour, I recommend checking response and raise exception if something went wrong.

File: ted_data_eu/services/data_load_service.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import json
+from typing import Dict, List
+
+from ted_data_eu.adapters.storage import ElasticStorage
+from ted_data_eu.adapters.storage_abc import DocumentStorageABC
+
+ELASTIC_DEFAULT_INDEX = 'ted_data'
```
Comment (CaptainOfHacks) [inline ted_data_eu/services/data_load_service.py:7]: add this variable in config

File: tests/e2e/test_document_storage.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+from ted_data_eu.services.data_load_service import load_documents_to_storage
+
+
+def test_elastic_storage_service(elastic_storage, document_file_path):
+    assert document_file_path.exists()
+    test_doc = json.loads(document_file_path.read_text(encoding='utf-8'))
+    result = load_documents_to_storage([test_doc, test_doc], elastic_storage)
+    assert result['errors'] is False
+    for item in result['items']:
+        assert item['create']['status'] == 201
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:12]: I recommend to implement Exception mechanism when something went wrong, and create custom Exceptions for Elastic HTTP API exceptions.

**Review by CaptainOfHacks — CHANGES_REQUESTED** [review]
Great implementation! Just some adjustments need to do.

File: ted_data_eu/adapters/storage.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            :param document: Document to be stored
+            :return: dict with Elastic API Response
+        """
+        response = requests.post(url=f"{self.host}/{self.elastic_index}/_doc/",
+                                 headers=BASIC_HEADERS,
+                                 auth=self.auth,
+                                 data=json.dumps(document))
+        deserialized_response = json.loads(response.content)
+        if response.status_code != 201:
+            raise Exception(
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:50]: Create custom exception, is difficult to catch general Exception, create ElasticStorageExaption class.

File: ted_data_eu/services/data_load_service.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+                              storage: DocumentStorageABC = None) -> Dict:
+    """
+        Loads data in document format to a storage.
+
+    :param documents: List of dict to be stored
+    :param storage: Document storage where Data will be loaded. If not specified, default (Elastic Storage) will be used
+    :return: dict with Elastic API Response
+    """
+
+    if not storage:
```
Comment (CaptainOfHacks) [inline ted_data_eu/services/data_load_service.py:18]: replace not storage to storage is None

File: ted_data_eu/adapters/storage.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                f"Elastic API add_document() error: {deserialized_response['error']['caused_by']['reason']}")
+        return deserialized_response
+
+    def add_documents(self, documents: List[Dict]) -> Dict:
```
Comment (CaptainOfHacks) [inline ted_data_eu/adapters/storage.py:54]: This method should return nothing, when we return JSON response we create a dependency where we call this function.

File: tests/e2e/test_document_storage.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import json
+
+from ted_data_eu.services.data_load_service import load_documents_to_storage
+
+
+def test_elastic_storage_service(elastic_storage, document_file_path):
+    assert document_file_path.exists()
+    test_doc = json.loads(document_file_path.read_text(encoding='utf-8'))
+    result = load_documents_to_storage([test_doc, test_doc], elastic_storage)
+    assert result['errors'] is False
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:10]: If we correctly decoupling, this type of verification should be removed.

File: tests/e2e/test_document_storage.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    for item in result['items']:
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:11]: Here I see, we need a Storage method to inspect an index, how many documents are in a specific index or get list of documents.

File: tests/e2e/test_document_storage.py (hunk 10 above)
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:12]: After we insert documents in storage, we can check with a get function.

Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:12]: Here I see an unknown for service layer response structure. We need to remove this response and use only raise Exception mechanism when something went wrong.

File: tests/e2e/test_document_storage.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        assert item['create']['status'] == 201
+
+
+def test_elastic_storage(elastic_storage, document_file_path, elastic_query):
+    assert document_file_path.exists()
+    test_doc = json.loads(document_file_path.read_text())
+    response = elastic_storage.add_document(test_doc)
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:18]: Here I see an unknown for service layer response structure. We need to remove this response and use only raise Exception mechanism when something went wrong.

File: tests/e2e/test_document_storage.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    assert response['result'] == 'created'
+
+    response = elastic_storage.query(elastic_query)
+    assert response['_shards']['successful'] > 0
+
+    response = elastic_storage.clear()
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:24]: Here I see an unknown for service layer response structure. We need to remove this response and use only raise Exception mechanism when something went wrong.

File: tests/e2e/test_document_storage.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def test_elastic_storage(elastic_storage, document_file_path, elastic_query):
+    assert document_file_path.exists()
+    test_doc = json.loads(document_file_path.read_text())
+    response = elastic_storage.add_document(test_doc)
+    assert response['result'] == 'created'
+
+    response = elastic_storage.query(elastic_query)
+    assert response['_shards']['successful'] > 0
```
Comment (CaptainOfHacks) [inline tests/e2e/test_document_storage.py:22]: for QueryResult we need a concrete ObjectType, this general dictionary response introduce an ElasticSearch direct dependency ( and a unknown dictionary structure ).

**Review by CaptainOfHacks — CHANGES_REQUESTED** [review]
Just a few little adjustments need to do.
The scope of an adapter is to full decoupling the internal system logic and external system's logic.
The scope of a service is to resolve/realize a concrete need (this need use only internal services, adapters, without external dependencies as logic, behaviour, errors, data structures, HTTP REST API response structure, or any uncontrolled in our system resources).

**Review by CaptainOfHacks — APPROVED** [review]

**Pull request merged**